funsor/terms.py

```diff
@@ -1057,53 +1058,95 @@ class Stack(Funsor):
-        self.components = components
+        super().__init__(inputs, output, fresh)
         self.name = name
+        self.parts = parts
 
     def eager_subs(self, subs):
         assert isinstance(subs, tuple) and len(subs) == 1 and subs[0][0] == self.name
         index = subs[0][1]
 
         # Try to eagerly select an index.
-        assert index.output == bint(len(self.components))
+        assert index.output == bint(len(self.parts))
 
         if isinstance(index, Number):
-            # Select a single component.
-            return self.components[index.data]
+            # Select a single part.
+            return self.parts[index.data]
         elif isinstance(index, Variable):
             # Rename the stacking dimension.
-            components = self.components
-            return Stack(components, index.name)
+            parts = self.parts
+            return Stack(index.name, parts)
         elif isinstance(index, Slice):
-            components = self.components[index.slice]
-            return Stack(components, index.name)
+            parts = self.parts[index.slice]
+            return Stack(index.name, parts)
         else:
             raise NotImplementedError('TODO support advanced indexing in Stack')
 
     def eager_reduce(self, op, reduced_vars):
-        components = self.components
+        parts = self.parts
         if self.name in reduced_vars:
             reduced_vars -= frozenset([self.name])
             if reduced_vars:
-                components = tuple(x.reduce(op, reduced_vars) for x in components)
-            return reduce(op, components)
-        components = tuple(x.reduce(op, reduced_vars) for x in components)
-        return Stack(components, self.name)
+                parts = tuple(x.reduce(op, reduced_vars) for x in parts)
+            return reduce(op, parts)
+        parts = tuple(x.reduce(op, reduced_vars) for x in parts)
+        return Stack(self.name, parts)
+
+
+class Cat(Funsor):
+    """
+    Concatenate funsors along an existing input dimension.
+
+    :param str name: The name of the input variable along which to concatenate.
+    :param tuple parts: A tuple of Funsors of homogenous output domain.
+    """
+    def __init__(self, name, parts):
+        assert isinstance(name, str)
+        assert isinstance(parts, tuple)
+        assert parts
+        assert all(name in x.inputs for x in parts)
+        assert len(set(x.output for x in parts)) == 1
+        output = parts[0].output
+        inputs = OrderedDict()
+        for x in parts:
+            inputs.update(x.inputs)
+        inputs[name] = bint(sum(x.inputs[name].size for x in parts))
+        super().__init__(inputs, output)
+        self.name = name
+        self.parts = parts
+
+
+@eager.register(Cat, str, tuple)
+def eager_cat(name, parts):
+    if len(parts) == 1:
+        return parts[0]
+    return eager_cat_homogeneous(name, *parts)
+
+
+@dispatch(str, Variadic[Funsor])
+def eager_cat_homogeneous(name, *parts):
```

Comment:
Why use this pattern instead of `eager.register(Cat, str, Variadic[Funsor])`?

Reply:
This is a weak aesthetic preference. It feels like we're compromising by using `multipledispatch` (thank you matt!) whereas funsor really deserves nested pattern matching. So I've made a nice tidy `Cat` interface with a nested datatype, and added this shim to deal with a library that can't handle nested patterns. Hopefully funsor will someday be reimplemented in a cleaner language than Python and this nested `Cat` datatype will survive.

But I'm happy to refactor to a flattened `Cat` signature if you think it's cleaner to hard code to current limitations of `multipledispatch`. If you are using the flattened pattern elsewhere in funsor, I guess we can be consistent and use that pattern here.